Bloodhound changes

rpkelly opened the pull request:
<!--
Tips:
- Please read CONTRIBUTING.md to understand our process and our requests for PRs.
- Please file an issue before creating a PR so we can discuss the change and confirm it's not already being worked on.
-->

**Issue number:**

Closes #540

**Description of changes:**
Adds the ability for bloodhound test overrides to exist, with reasoning, and optional messaging in the case of skipped tests. Additionally, overrides 2 tests:
* Bottlerocket 03040101 if running on kubernetes variants will now be overridden to skip with an explanation
* Kubernetes 04021000 was already set to automatically pass, but has been switched to use the new override system

Finally, we've removed the RSA ciphers no longer considered secure by the latest kubernetes CIS benchmark from k8s 04021200

**Testing done:**

Run tests:
```
[ssm-user@control]$ apiclient exec report cis -l 2
ctr: container "report" in namespace "default": not found
[ssm-user@control]$ apiclient report cis -l 2
Benchmark name:  CIS Bottlerocket Benchmark
Version:         v1.0.0
Reference:       https://www.cisecurity.org/benchmark/bottlerocket
Benchmark level: 2
Start time:      2025-10-20T22:08:02.530202650Z

[FAIL] 1.1.1.1   Ensure mounting of udf filesystems is disabled (Automatic)
[SKIP] 1.2.1     Ensure software update repositories are configured (Manual)
[PASS] 1.3.1     Ensure dm-verity is configured (Automatic)
[PASS] 1.4.1     Ensure setuid programs do not create core dumps (Automatic)
[PASS] 1.4.2     Ensure address space layout randomization (ASLR) is enabled (Automatic)
[PASS] 1.4.3     Ensure unprivileged eBPF is disabled (Automatic)
[PASS] 1.4.4     Ensure user namespaces are disabled (Automatic)
[PASS] 1.5.1     Ensure SELinux is configured (Automatic)
[PASS] 1.5.2     Ensure Lockdown is configured (Automatic)
[SKIP] 1.6       Ensure updates, patches, and additional security software are installed (Manual)
[PASS] 2.1.1.1   Ensure chrony is configured (Automatic)
[FAIL] 3.1.1     Ensure packet redirect sending is disabled (Automatic)
[PASS] 3.2.1     Ensure source routed packets are not accepted (Automatic)
[FAIL] 3.2.2     Ensure ICMP redirects are not accepted (Automatic)
[FAIL] 3.2.3     Ensure secure ICMP redirects are not accepted (Automatic)
[FAIL] 3.2.4     Ensure suspicious packets are logged (Automatic)
[PASS] 3.2.5     Ensure broadcast ICMP requests are ignored (Automatic)
[PASS] 3.2.6     Ensure bogus ICMP responses are ignored (Automatic)
[PASS] 3.2.7     Ensure TCP SYN Cookies is enabled (Automatic)
[FAIL] 3.3.1     Ensure SCTP is disabled (Automatic)
[SKIP] 3.4.1.1   Ensure IPv4 default deny firewall policy (Exception)
[FAIL] 3.4.1.2   Ensure IPv4 loopback traffic is configured (Automatic)
[SKIP] 3.4.1.3   Ensure IPv4 outbound and established connections are configured (Manual)
[FAIL] 3.4.2.1   Ensure IPv6 default deny firewall policy (Automatic)
[FAIL] 3.4.2.2   Ensure IPv6 loopback traffic is configured (Automatic)
[SKIP] 3.4.2.3   Ensure IPv6 outbound and established connections are configured (Manual)
[PASS] 4.1.1.1   Ensure journald is configured to write logs to persistent disk (Automatic)
[PASS] 4.1.2     Ensure permissions on journal files are configured (Automatic)

Passed:          14
Failed:          9
Skipped:         5
Total checks:    28

Compliance check result: FAIL

Exceptions:
*  3.4.1.1 Test requires iptables FORWARD chain configuration that conflicts with Kubernetes networking architecture. Kubernetes manages its own iptables rules through kube-proxy and CNI plugins.
[ssm-user@control]$ apiclient report cis-k8s -l 2
Benchmark name:  CIS Kubernetes Benchmark (Worker Node)
Version:         v1.11.1
Reference:       https://www.cisecurity.org/benchmark/kubernetes
Benchmark level: 2
Start time:      2025-10-20T22:08:10.965751661Z

[PASS] 4.1.1     Ensure that the kubelet service file permissions are set to 600 or more restrictive (Automatic)
[PASS] 4.1.2     Ensure that the kubelet service file ownership is set to root:root (Automatic)
[SKIP] 4.1.3     If proxy kubeconfig file exists ensure permissions are set to 600 or more restrictive (Manual)
[SKIP] 4.1.4     If proxy kubeconfig file exists ensure ownership is set to root:root (Manual)
[PASS] 4.1.5     Ensure that the --kubeconfig kubelet.conf file permissions are set to 600 or more restrictive (Automatic)
[PASS] 4.1.6     Ensure that the --kubeconfig kubelet.conf file ownership is set to root:root (Automatic)
[PASS] 4.1.7     Ensure that the certificate authorities file permissions are set to 644 or more restrictive (Automatic)
[PASS] 4.1.8     Ensure that the client certificate authorities file ownership is set to root:root (Automatic)
[PASS] 4.1.9     If the kubelet config.yaml configuration file is being used validate permissions set to 600 or more restrictive (Automatic)
[PASS] 4.1.10    If the kubelet config.yaml configuration file is being used validate file ownership is set to root:root (Automatic)
[PASS] 4.2.1     Ensure that the --anonymous-auth argument is set to false (Automatic)
[PASS] 4.2.2     Ensure that the --authorization-mode argument is not set to AlwaysAllow (Automatic)
[PASS] 4.2.3     Ensure that the --client-ca-file argument is set as appropriate (Automatic)
[PASS] 4.2.4     Verify that if defined, readOnlyPort is set to 0 (Automatic)
[PASS] 4.2.5     Ensure that the --streaming-connection-idle-timeout argument is not set to 0 (Automatic)
[PASS] 4.2.6     Ensure that the --make-iptables-util-chains argument is set to true (Automatic)
[SKIP] 4.2.7     Ensure that the --hostname-override argument is not set (not valid for Bottlerocket) (Manual)
[SKIP] 4.2.8     Ensure that the eventRecordQPS argument is set to a level which ensures appropriate event capture (Manual)
[PASS] 4.2.9     Ensure that the --tls-cert-file and --tls-private-key-file arguments are set as appropriate (Automatic)
[PASS] 4.2.10    Ensure that the --rotate-certificates argument is not set to false (Exception)
[PASS] 4.2.11    Verify that the RotateKubeletServerCertificate argument is set to true (Automatic)
[PASS] 4.2.12    Ensure that the Kubelet only makes use of Strong Cryptographic Ciphers (Automatic)
[PASS] 4.2.13    Ensure that a limit is set on pod PIDs (Automatic)
[FAIL] 4.2.14    Ensure that the --seccomp-default parameter is set to true (Automatic)
[SKIP] 4.2.15    Ensure that the --IPAddressDeny is set to any (Manual)
[SKIP] 4.3.1     Ensure that the kube-proxy metrics service is bound to localhost (Manual)

Passed:          19
Failed:          1
Skipped:         6
Total checks:    26

Compliance check result: FAIL

Exceptions:
*  4.2.10 IAM (external) auth is used on Bottlerocket, so certificate rotation does not apply. See EKS CIS Benchmark.
```

Check to ensure overrides are on kubernetes variants, but not ECS variants:


```
ec2-user /mnt   20:31  ❯ cd bottlerocket-k8s-1.33/usr/libexec/cis-checks/

ec2-user …/usr/libexec/cis-checks   20:31  ❯ ls
bottlerocket  kubernetes

ec2-user …/usr/libexec/cis-checks   20:31  ❯ ls bottlerocket/
br01010101  br01030100  br01040200  br01040400  br01050200  br02010101  br03020100  br03020300  br03020500  br03020700  br03040101       br03040102  br03040201  br03040203  br04010200
br01020100  br01040100  br01040300  br01050100  br01060000  br03010100  br03020200  br03020400  br03020600  br03030100  br03040101.json  br03040103  br03040202  br04010101  metadata.json

ec2-user …/usr/libexec/cis-checks   20:31  ❯ ls kubernetes/
k8s04010100  k8s04010300  k8s04010500  k8s04010700  k8s04010900  k8s04020100  k8s04020300  k8s04020500  k8s04020700  k8s04020900  k8s04021000.json  k8s04021200  k8s04021400  k8s04030100
k8s04010200  k8s04010400  k8s04010600  k8s04010800  k8s04011000  k8s04020200  k8s04020400  k8s04020600  k8s04020800  k8s04021000  k8s04021100       k8s04021300  k8s04021500  metadata.json

ec2-user …/usr/libexec/cis-checks   20:31  ❯ ....

ec2-user /mnt/bottlerocket-k8s-1.33   20:31  ❯ ..

ec2-user /mnt   20:31  ❯ ls
bottlerocket-ecs-2  bottlerocket-k8s-1.33  bottlerocket-root

ec2-user /mnt   20:31  ❯ cd bottlerocket-ecs-2/usr/libexec/cis-checks/

ec2-user …/usr/libexec/cis-checks   20:31  ❯ ls
bottlerocket

ec2-user …/usr/libexec/cis-checks   20:31  ❯ ls bottlerocket/
br01010101  br01030100  br01040200  br01040400  br01050200  br02010101  br03020100  br03020300  br03020500  br03020700  br03040101  br03040103  br03040202  br04010101  metadata.json
br01020100  br01040100  br01040300  br01050100  br01060000  br03010100  br03020200  br03020400  br03020600  br03030100  br03040102  br03040201  br03040203  br04010200
```

json output sample: 

```
   "error": ""
    },
    "br03030100": {
      "name": "br03030100",
      "id": "3.3.1",
      "level": 2,
      "title": "Ensure SCTP is disabled",
      "mode": "Automatic",
      "status": "FAIL",
      "error": "modprobe for sctp is not disabled"
    },
    "br03040101": {
      "name": "br03040101",
      "id": "3.4.1.1",
      "level": 2,
      "title": "Ensure IPv4 default deny firewall policy",
      "mode": "Automatic",
      "status": "SKIP",
      "error": "Test requires iptables FORWARD chain configuration that conflicts with Kubernetes networking architecture. Kubernetes manages its own iptables rules through kube-proxy and CNI plugins."
    },
    "br03040102": {
      "name": "br03040102",
      "id": "3.4.1.2",
      "level": 2,
      "title": "Ensure IPv4 loopback traffic is configured",
      "mode": "Automatic",
      "status": "FAIL",
      "error": "Unable to find expected iptables INPUT values"
    },
    "br03040103": {
      "name": "br03040103",
      "id": "3.4.1.3",
```

**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

Commit:
bloodhound: activate k8s04021000 checker and add override json

Replace ManualChecker with actual K8S04021000Checker implementation
and add override to pass on Bottlerocket.

Signed-off-by: Richard Kelly <rpkelly@amazon.com>

Commit:
bloodhound: remove RSA ciphers from k8s04021200 allowed list

Remove TLS_RSA_WITH_AES_128_GCM_SHA256 and TLS_RSA_WITH_AES_256_GCM_SHA384
from allowed cipher suites per latest Kubernetes CIS benchmark.

Signed-off-by: Richard Kelly <rpkelly@amazon.com>

Commit:
bloodhound: implement test override system

Add override system to skip specific compliance tests through JSON
configuration files. Enables variants to skip incompatible
tests while maintaining compliance checking.

Signed-off-by: Richard Kelly <rpkelly@amazon.com>

ytsssun commented:
Q - could you explain on the failed tests in `apiclient report -l 2 cis` and are they expected?

rpkelly commented:
@ytsssun The fails are expected. I did not set the image to be L2 CIS compliant, I only wanted to validate that the overrides were being properly read.

bcressey commented:
The formatting for this line is wonky in plain text output:
```
[SKIPPED: Test requires iptables FORWARD chain configuration that conflicts with Kubernetes networking architecture. Kubernetes manages its own iptables rules through kube-proxy and CNI plugins.] 3.4.1.1   Ensure IPv4 default deny firewall policy (Automatic)
```

Ideally we'd format it more like:
```
[SKIP] 3.4.1.1   Ensure IPv4 default deny firewall policy (Exception)

Exceptions:
*  3.4.1.1 requires iptables FORWARD chain configuration that conflicts with Kubernetes networking architecture. Kubernetes manages its own iptables rules through kube-proxy and CNI plugins.
```

bcressey commented on sources/bloodhound/src/results.rs:
```diff
@@ -22,6 +22,8 @@ pub enum CheckStatus {
     #[default]
     SKIP,
+    /// Test was skipped due to override configuration.
+    SKIPPEDOVERRIDE,
```
nit: maybe just `OVERRIDE` here.

bcressey commented on packages/os/br03040101.json:
```diff
@@ -0,0 +1,16 @@
+{
+  "version": "1.0",
```
Is this a schema version or how is it meant to be used? By humans or programmatically?

rpkelly replied:
It's meant to be a schema version, but, given the below feedback concerning the metadata object, I can change the entire file to just be

```
{
   "reason":
   "reference":
   "status":
}
```

bcressey commented on packages/os/br03040101.json:
```diff
@@ -0,0 +1,16 @@
+  "metadata": {
+    "description": "Test overrides for Kubernetes variants of Bottlerocket",
+    "created_date": "2025-01-09",
+    "applies_to": ["aws-k8s"]
```
Do we need this metadata? What information would be lost if it's removed?

`applies_to` can be inferred from the packaging, `created_date` from Git history, and the description seems like it will just be a less-specific summary of the reasons.

bcressey commented on packages/os/br03040101.json:
```diff
@@ -0,0 +1,16 @@
+  },
+  "overrides": [
+    {
+      "test_id": "br03040101",
```
You're establishing one convention with the file name - which implies one override per file - and a different one with the file format, which ignores the file name.

I would say pick one or the other - either enforce a 1:1 match between file names and test IDs, or (easier) just name the file something else, like `bottlerocket-cis-checks-aws-k8s-overrides.json`.

bcressey commented on packages/os/br03040101.json:
```diff
@@ -0,0 +1,16 @@
+    {
+      "test_id": "br03040101",
+      "reason": "Test requires iptables FORWARD chain configuration that conflicts with Kubernetes networking architecture. Kubernetes manages its own iptables rules through kube-proxy and CNI plugins.",
+      "reference": "https://github.com/bottlerocket-os/bottlerocket/issues/12345",
```
Is this just a placeholder value?

bcressey commented on sources/bloodhound/src/main.rs:
```diff
@@ -54,62 +55,146 @@ fn read_metadata(check_dir: &Path) -> ReportMetadata {
+        "FAIL" => CheckStatus::FAIL,
+        "SKIP" => CheckStatus::SKIPPEDOVERRIDE,
+        _ => CheckStatus::SKIPPEDOVERRIDE, // Default to skip for unknown statuses
+    };
```
It's kind of strange to pass-through some statuses and only special-case SKIP.

This might be a terrible idea but you could try a self-referential enum that wraps one of the statuses in an override status, so we don't discard that information:
```
enum CheckStatus {
    PASS,
    FAIL,
    SKIP,
    OVERRIDE(Box<CheckStatus>),
}
```

bcressey commented on sources/bloodhound/src/main.rs:
```diff
@@ -54,62 +55,146 @@ fn read_metadata(check_dir: &Path) -> ReportMetadata {
+fn find_overrides(check_dir: &Path, test_names: &[String]) -> HashMap<String, OverrideConfig> {
+    let mut overrides = HashMap::new();
+    for test_name in test_names {
+        let override_path = check_dir.join(format!("{test_name}.json"));
+        if override_path.exists() {
+            match load_override_config(&override_path) {
+                Ok(override_config) => {
+                    overrides.insert(test_name.clone(), override_config);
+                }
+                Err(err) => {
+                    eprintln!("Warning: Failed to load override file {override_path:?}: {err}");
+                }
```
Per the other note - we should either check the filename or check that the file has a matching test name, but not both.

rpkelly commented:
Added updated code, will rebuild and run a test image to show the new output next

rpkelly commented:
Updated test output has been added

bcressey commented on packages/os/br03040101.json:
```diff
@@ -0,0 +1,5 @@
+{
+  "reason": "Test requires iptables FORWARD chain configuration that conflicts with Kubernetes networking architecture. Kubernetes manages its own iptables rules through kube-proxy and CNI plugins.",
+  "reference": "https://github.com/bottlerocket-os/bottlerocket/issues/540",
+  "status": "SKIP"
+}
```
Do we also need an override for the IPv6 check, 3.4.2.1?

rpkelly replied:
Per https://github.com/bottlerocket-os/bottlerocket-core-kit/issues/540 3.4.2.1 actually _does_ pass if properly configured, as we only seem to override the IPv4 rules.

bcressey replied:
> Per https://github.com/bottlerocket-os/bottlerocket-core-kit/issues/540 3.4.2.1 actually does pass if properly configured, as we only seem to override the IPv4 rules.

But does the node actually work properly if using IPv6 for the overlay network, with forwarding disabled? I'm not sure why that would be true for IPv6 but not true for IPv4.

bcressey commented on sources/bloodhound/src/output.rs:
```diff
@@ -56,3 +78,156 @@ impl ReportWriter for JsonReportWriter {
+        // Verify Exceptions section contains both reasons
+        assert!(output_str.contains("Exceptions:"));
+        assert!(output_str.contains("*  1.1.1 Override reason for pass"));
+        assert!(output_str.contains("*  2.2.2 Override reason for fail"));
```
Looking at this output, it seems like overriding to a pass vs. a failure is indistinguishable. Maybe we should just insert the override value into the original bracketed text, so this would become:
```
[PASS] 1.1.1     Override pass test (Exception)
[FAIL] 2.2.2     Override fail test (Exception)
[SKIP] 3.3.3     Override skip test (Exception)

Exceptions:
*  1.1.1 Override reason for pass
*  2.2.2 Override reason for fail
*  3.3.3 Override reason for skip
```

bcressey commented on sources/bloodhound/src/output.rs:
```diff
@@ -56,3 +78,156 @@ impl ReportWriter for JsonReportWriter {
+    fn test_text_output_override_pass_fail_statuses() {
+        // GIVEN: A report with overrides that have PASS and FAIL statuses
+        // WHEN: TextReportWriter writes the report
+        // THEN: Overrides should show original status but still appear in Exceptions section
```
This doesn't seem to be implemented, unless I'm misunderstanding what "show original status" means.

bcressey commented on sources/bloodhound/src/output.rs:
```diff
@@ -56,3 +78,156 @@ impl ReportWriter for JsonReportWriter {
+    fn test_text_output_multiple_exceptions() {
+        // GIVEN: A report with multiple overridden tests and regular tests
+        // WHEN: TextReportWriter writes the report
+        // THEN: Exceptions section should list all override reasons
```
This seems so close to the other test that it might be best to merge them.

bcressey commented on sources/bloodhound/src/main.rs:
```diff
@@ -54,62 +54,137 @@ fn read_metadata(check_dir: &Path) -> ReportMetadata {
+            error: String::new(),
+            override_reason: Some(override_config.reason.clone()),
+        },
+    );
```
`original_status` seems like a misnomer - it's just the status from the json file, it's not the result we would have gotten if no override were present

rpkelly replied:
Changing it to `override_status`

bcressey commented on sources/bloodhound/src/output.rs:
```diff
@@ -26,14 +26,26 @@ impl ReportWriter for TextReportWriter {
+                        exceptions.push((test_result.metadata.id.clone(), reason.clone()));
+                    }
+                    ("SKIP".to_string(), "Exception".to_string())
+                }
```
Instead of hard-coding SKIP, we can take the embedded `CheckStatus` value here.

bcressey commented on sources/bloodhound/src/results.rs:
```diff
@@ -174,7 +191,7 @@ impl ReportResults {
                 }
             }
-            CheckStatus::SKIP => {
+            CheckStatus::SKIP | CheckStatus::OVERRIDE(_) => {
```
Here I think we'd increment the counts based on the override status, e.g. `CheckStatus::PASS` or `CheckStatus::OVERRIDE(CheckStatus::PASS)` would increment `self.passed += 1`.

bcressey reviewed (commented):
This is very close, but it seems like it'd be nice if the implementation respected the status value provided in the override file when rendering the report.

bcressey commented on packages/os/os.spec:
```diff
@@ -407,6 +410,12 @@ Requires: (%{_cross_os}bloodhound and %{_cross_os}variant-runtime(k8s))
 %description -n %{_cross_os}bloodhound-k8s
 %{summary}.
 
+%package -n %{_cross_os}bloodhound-k8s-overrides
+Summary: CIS Test overrides for Kubernetes variants
+Requires: (%{_cross_os}bloodhound and %{_cross_os}variant-runtime(k8s))
+%description -n %{_cross_os}bloodhound-k8s-overrides
+%{summary}.
```
nit: can just add these into the existing `bloodhound-k8s` subpackage which has the same requirements

bcressey reviewed (commented):
How are exceptions shown in the JSON output?

Commit:
bloodhound: maintain json format compatibily

The added test overrides do not directly flatten into the previous json
output structure. In order to maintain the same json structure we
require a custom serializer

Signed-off-by: Richard Kelly <rpkelly@amazon.com>
